homeassistant/components/minio/__init__.py

```diff
@@ -0,0 +1,282 @@
+            EVENT_HOMEASSISTANT_START,
+            minio_listener.start_handler
+        )
+        hass.bus.listen_once(
+            EVENT_HOMEASSISTANT_STOP,
+            minio_listener.stop_handler
+        )
+
+    for listen_conf in conf[CONF_LISTEN]:
+        _setup_listener(listen_conf)
+
+    minio_client = Minio(
+        get_minio_endpoint(host, port),
+        access_key,
+        secret_key,
+        secure
+    )
+
+    def _render_service_value(service, key):
+        value = service.data[key]
+        value.hass = hass
+        return value.async_render()
+
+    def put_file(service):
+        """Upload file service."""
+        bucket = _render_service_value(service, 'bucket')
+        key = _render_service_value(service, 'key')
+        file_path = _render_service_value(service, 'file_path')
+
+        if not hass.config.is_allowed_path(file_path):
+            _LOGGER.error('Invalid file_path %s', file_path)
+            return
+
+        minio_client.fput_object(bucket, key, file_path)
+
+    def get_file(service):
+        """Download file service."""
+        bucket = _render_service_value(service, 'bucket')
+        key = _render_service_value(service, 'key')
+        file_path = _render_service_value(service, 'file_path')
+
+        if not hass.config.is_allowed_path(file_path):
+            _LOGGER.error('Invalid file_path %s', file_path)
+            return
+
+        minio_client.fget_object(bucket, key, file_path)
+
+    def remove_file(service):
+        """Delete file service."""
+        bucket = _render_service_value(service, 'bucket')
+        key = _render_service_value(service, 'key')
+
+        minio_client.remove_object(bucket, key)
+
+    hass.services.register(
+        DOMAIN, 'put', put_file, schema=BUCKET_KEY_FILE_SCHEMA
+    )
+    hass.services.register(
+        DOMAIN, 'get', get_file, schema=BUCKET_KEY_FILE_SCHEMA
+    )
+    hass.services.register(
+        DOMAIN, 'remove', remove_file, schema=BUCKET_KEY_SCHEMA
+    )
+
+    return True
+
+
+def get_minio_endpoint(host: str, port: int) -> str:
+    """Create minio endpoint from host and port."""
+    return "{}:{}".format(host, port)
+
+
+class QueueListener(threading.Thread):
+    """Forward events from queue into HASS event bus."""
+
+    def __init__(self, hass):
+        """Create queue."""
+        super().__init__()
+        self._hass = hass
+        self._queue = Queue()
+
+    def run(self):
+        """Listen to queue events, and forward them to HASS event bus."""
+        _LOGGER.info('Running QueueListener')
+        while True:
+            event = self._queue.get()
+            if event is None:
+                break
+
+            _, file_name = os.path.split(event['key'])
+
+            _LOGGER.debug(
+                'Sending event %s, %s, %s',
+                event['event_name'],
+                event['bucket'],
+                event['key']
+            )
+            self._hass.bus.async_fire(DOMAIN, {
```

Comment:
Don't use the async api. That can only be used inside the event loop thread. Here we're in a separate thread.

Reply:
But for some reason the test fails when using sync `.fire()` method. As the `await hass.async_block_till_done()` doesn't wait for the listen callback to be called

Reply:
We have to solve that in the test.